SparseGrids/tsgGridLocalPolynomial.hpp

```diff
@@ -181,6 +186,8 @@ class GridLocalPolynomial : public BaseCanonicalGrid{
      * - \b mode \b 0, ignore \b sindx and \b svals, find the non-zero basis functions multiply them by the surpluses and add to \b y
      * - \b mode \b 1, ignore \b y, form a sparse vector by std::vector::push_back() to the \b sindx and \b svals
      * - \b mode \b 2, same as \b mode \b 1 but it also sorts the entries within the vector (requirement of Nvidia cusparseDgemvi)
+     * - \b mode \b 3, same as \b mode \b 0, but replaces the non-zero basis function values with their derivatives
+     * - \b mode \b 4, same as \b mode \b 1, but replaces the non-zero basis function values with their derivatives
```

Comment:
Do you need mode 4? We are not building a sparse matrix for the derivatives, this is used only in the case where we are computing the derivatives in batches and that's a whole other can of worms.

Reply:
Mode 4 is needed in `GridLocalPolynomial::getDifferentiationWeights()`, which does solve a large, but sparse, triangular linear system of equations.

Reply:
I don't get it. The matrix that should be used here is the one from mode 1, there is no special separate matrix for the derivatives.

Reply:
I'm not sure about that. Since we have `num_dimension` more variables to solve from, shouldn't the matrix have `num_dimension` times more entries than the one for mode 1? Also, shouldn't it be filled with the derivative values rather than the function values?

Reply:
You would only ever form the matrix here in the case of batch differentiation. We don't differentiate in batch.

The only matrix that we are working with is the one corresponding to the computation of surplus coefficients and that matrix does not depend on `x` and thus is a constant with respect to differentiation.

Reply:
Remove case 4 and I think this is ready.

Reply:
Maybe I am missing something here conceptually. 

Suppose a real-valued interpolant `F` is given by `F(x) = <Φ(x), s>`, where `Φ(x)` is vector of basis functions values evaluated at `x` and `s` is the vector of surplus coefficients. As far as I can tell, `walkTree<1>()` is generating `Φ(x)` and directly depends on `x`. My implementation of `walkTree<4>()` is just generating `∇Φ(x)`, which clearly cannot be the same matrix/vector as `Φ(x)`.

Note that the application of the inverse transpose of the matrix associated with the surplus coefficients is done in the calls of `applyTransformationTransposed<T>()` and has nothing to do with `walkTree<T>()`.